src/jobservice/job/impl/gc/garbage_collection.go

```diff
@@ -186,6 +114,162 @@ func (gc *GarbageCollector) init(ctx job.Context, params job.Parameters) error {
 			gc.deleteUntagged = untagged
 		}
 	}
+	// default is not execute dry run, and for dry run we can have button in the UI.
+	dryRun, exist := params["dry_run"]
+	if exist {
+		if dryRun, ok := dryRun.(bool); ok && !dryRun {
+			gc.dryRun = dryRun
+		}
+	}
+	return nil
+}
+
+// Run implements the interface in job/Interface
+func (gc *GarbageCollector) Run(ctx job.Context, params job.Parameters) error {
+	if err := gc.init(ctx, params); err != nil {
+		return err
+	}
+
+	gc.logger.Infof("start to run gc in job.")
+
+	// mark
+	if err := gc.mark(ctx); err != nil {
+		gc.logger.Errorf("failed to execute GC job at mark phase, error: %v", err)
+		return err
+	}
+
+	// sweep
+	if !gc.dryRun {
+		if err := gc.sweep(ctx); err != nil {
+			gc.logger.Errorf("failed to execute GC job at sweep phase, error: %v", err)
+			return err
+		}
+
+		if err := gc.cleanCache(); err != nil {
+			return err
+		}
+	}
+	gc.logger.Infof("success to run gc in job.")
+	return nil
+}
+
+// mark
+func (gc *GarbageCollector) mark(ctx job.Context) error {
+	arts, err := gc.deletedArt(ctx)
+	if err != nil {
+		gc.logger.Errorf("failed to delete GC candidates in gc job, with error: %v", err)
+		return err
+	}
+	// no need to execute GC as there is no removed artifacts.
+	// Do this is to handle if user trigger GC job several times, only one job should do the following logic as artifact trash table is flushed.
+	if len(arts) == 0 {
+		gc.logger.Info("no need to execute GC as there is no removed artifacts.")
+		return nil
+	}
```

Comment:
Upgrade from v1.9 and v1.10, the ArtifactTrash table is empty but some blobs may need to delete in GC.

Reply:
I have considered about it, but if the user triggers the GC for the first time without delete any of artifact, GC will delete some blobs, it seemed a little strange to him. 

For the undeleted blobs in the older version, once user gets some artifat deleted, then the GC job will remove them.

Reply:
GC happens not because users had deleted the artifacts it happens because there are blobs that need to delete.

Reply:
Oh, I know that when then ArtifactTrash is empty, we can't get the repository of the blob, it's so sad.